### ISSUE TITLE
Unable to use 'ddp' strategy: "No such file or directory" on initializing distributed

### ISSUE THREAD

**Issue by esaracin** [Lightning-AI/lightning#17530]
### Bug description

I'm unable to use the suggested strategy of 'ddp' when training my model across multiple (4) GPUs on a single node. Using the default strategy of ddp_spawn leads to bugs and doesn't seem very stable (which is what the documentation seems to suggest).

Specifically, when I use 'ddp' (or some variation of 'ddp') as my strategy, training always hangs on the .fit() call, as the processes are being created. The first of 4 distributed processes appears to be created, and the other 3 error out on a "No such file or directory" error, pointing at the main script being executed. It's worth noting that the program doesn't crash, it just hangs (and in fact I need to close the terminal/execute pkill to end it), which suggests some sort of deadlock.

<img width="1348" alt="Screenshot 2023-04-29 at 11 06 18 AM" src="https://user-images.githubusercontent.com/25374730/235311728-adbca010-1e06-4556-814d-759483cd2e1b.png">

This happened both on 1.9.5, and now on 2.0.2

### What version are you seeing the problem on?

v2_0

### How to reproduce the bug

```python
trainer = Trainer(
            deterministic=True,
            gradient_clip_val=0.5,
            precision=16,
            gradient_clip_algorithm="value",
            callbacks=callbacks,
            profiler="simple",
            max_epochs=self.max_epochs,
            logger=logger,
            devices=4,
            accelerator='gpu',
            strategy='ddp'
        )
trainer.fit(model, datamodule=datamodule)


Simply removing the strategy argument allows training to proceed as expected (when printing the strategy after omitting 'ddp', I've confirmed it defaults to DDPSpawn).
```

### Error messages and logs

Per the above screenshot, I get:
```
[Errno 2] No such file or directory for the 3 alternate processes during their initialization.
```

### Environment

* CUDA:
        - GPU:
                - Tesla M60
                - Tesla M60
                - Tesla M60
                - Tesla M60
        - available:         True
        - version:           10.2
* Lightning:
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.2
        - torch:             1.12.1
        - torchmetrics:      0.11.4
* Packages:
        - absl-py:           1.4.0
        - adjustment-model-data-pull: 0.0.5
        - aiobotocore:       2.3.4
        - aiohttp:           3.8.4
        - aioitertools:      0.11.0
        - aiosignal:         1.3.1
        - alembic:           1.10.4
        - anyio:             3.6.2
        - apache-airflow:    2.5.3
        - apache-airflow-providers-common-sql: 1.4.0
        - apache-airflow-providers-ftp: 3.3.1
        - apache-airflow-providers-http: 4.3.0
        - apache-airflow-providers-imap: 3.1.1
        - apache-airflow-providers-sqlite: 3.3.2
        - apispec:           3.3.2
        - argcomplete:       3.0.8
        - asgiref:           3.6.0
        - astunparse:        1.6.3
        - async-timeout:     4.0.2
        - attrs:             23.1.0
        - babel:             2.12.1
        - bandit:            1.7.5
        - beautifulsoup4:    4.12.2
        - black:             22.12.0
        - blinker:           1.6.2
        - blis:              0.7.9
        - boto3:             1.21.21
        - boto3-stubs:       1.26.123
        - botocore:          1.24.21
        - botocore-stubs:    1.29.123
        - cachelib:          0.9.0
        - cachetools:        5.3.0
        - catalogue:         2.0.8
        - cattrs:            22.2.0
        - certifi:           2022.12.7
        - cffi:              1.15.1
        - cfgv:              3.3.1
        - charset-normalizer: 3.1.0
        - click:             8.1.3
        - clickclick:        20.10.2
        - cloudpickle:       2.2.1
        - cmaes:             0.9.1
        - colorama:          0.4.6
        - colorlog:          4.8.0
        - confection:        0.0.4
        - configupdater:     3.1.1
        - connexion:         2.14.1
        - contourpy:         1.0.7
        - coverage:          6.5.0
        - cramjam:           2.6.2
        - cron-descriptor:   1.2.35
        - croniter:          1.3.14
        - cryptography:      40.0.2
        - cycler:            0.11.0
        - cymem:             2.0.7
        - cython:            0.29.34
        - dash:              2.9.3
        - dash-core-components: 2.0.0
        - dash-html-components: 2.0.0
        - dash-table:        5.0.0
        - databricks-cli:    0.17.6
        - dataclasses-json:  0.5.7
        - deprecated:        1.2.13
        - dill:              0.3.6
        - distlib:           0.3.6
        - dnspython:         2.3.0
        - docker:            6.0.1
        - docutils:          0.19
        - dunamai:           1.16.0
        - email-validator:   1.3.1
        - entrypoints:       0.4
        - et-xmlfile:        1.1.0
        - exceptiongroup:    1.1.1
        - execnet:           1.9.0
        - fastparquet:       0.7.2
        - filelock:          3.12.0
        - flake518:          1.5.0
        - flake8:            6.0.0
        - flask:             2.2.4
        - flask-appbuilder:  4.1.4
        - flask-babel:       2.0.0
        - flask-caching:     2.0.2
        - flask-jwt-extended: 4.4.4
        - flask-login:       0.6.2
        - flask-session:     0.4.0
        - flask-sqlalchemy:  2.5.1
        - flask-wtf:         1.1.1
        - flatbuffers:       23.3.3
        - fonttools:         4.39.3
        - frozenlist:        1.3.3
        - fsspec:            2022.7.1
        - gast:              0.4.0
        - gdown:             4.7.1
        - gensim:            4.3.1
        - gitdb:             4.0.10
        - gitpython:         3.1.31
        - google-auth:       2.17.3
        - google-auth-oauthlib: 0.4.6
        - google-pasta:      0.2.0
        - graphviz:          0.20.1
        - greenlet:          2.0.2
        - grpcio:            1.54.0
        - gunicorn:          20.1.0
        - h11:               0.14.0
        - h5py:              3.8.0
        - httpcore:          0.17.0
        - httpx:             0.24.0
        - identify:          2.5.23
        - idna:              3.4
        - imbalanced-learn:  0.10.1
        - imblearn:          0.0
        - importlib-metadata: 5.2.0
        - importlib-resources: 5.12.0
        - inflection:        0.5.1
        - iniconfig:         2.0.0
        - isort:             5.12.0
        - itsdangerous:      2.1.2
        - jax:               0.4.8
        - jinja2:            3.1.2
        - jmespath:          1.0.1
        - joblib:            1.2.0
        - jsonschema:        4.17.3
        - keras:             2.11.0
        - keras-preprocessing: 1.1.2
        - kiwisolver:        1.4.4
        - labs-continuous-learning-pipeline: 0.0.0
        - langcodes:         3.3.0
        - lazy-object-proxy: 1.9.0
        - legal-structured-transformers: 0.2.3
        - legal-text-transformers: 0.2.3
        - libclang:          16.0.0
        - lightning-utilities: 0.8.0
        - linkify-it-py:     2.0.0
        - lit-ds-utils:      3.6.0
        - lockfile:          0.12.2
        - luigi:             3.2.1
        - mako:              1.2.4
        - markdown:          3.4.3
        - markdown-it-py:    2.2.0
        - markupsafe:        2.1.2
        - marshmallow:       3.19.0
        - marshmallow-enum:  1.5.1
        - marshmallow-oneofschema: 3.0.1
        - marshmallow-sqlalchemy: 0.26.1
        - matplotlib:        3.7.1
        - mccabe:            0.7.0
        - mdit-py-plugins:   0.3.5
        - mdurl:             0.1.2
        - ml-dtypes:         0.1.0
        - mlflow:            1.30.1
        - mock:              4.0.3
        - multidict:         6.0.4
        - murmurhash:        1.0.9
        - mypy:              1.2.0
        - mypy-boto3-kms:    1.26.81
        - mypy-boto3-s3:     1.26.116
        - mypy-extensions:   1.0.0
        - nlpaug:            1.1.11
        - nltk:              3.8.1
        - nodeenv:           1.7.0
        - numpy:             1.24.1
        - oauthlib:          3.2.2
        - openpyxl:          3.1.2
        - opt-einsum:        3.3.0
        - optuna:            3.1.1
        - packaging:         21.3
        - pandas:            1.5.3
        - pandasql:          0.7.3
        - pathspec:          0.9.0
        - pathy:             0.10.1
        - pbr:               5.11.1
        - pendulum:          2.1.2
        - pillow:            9.5.0
        - pip:               23.1
        - platformdirs:      3.5.0
        - plotly:            5.14.1
        - pluggy:            1.0.0
        - poetry-dynamic-versioning: 0.14.1
        - pre-commit:        2.20.0
        - preshed:           3.0.8
        - prison:            0.2.1
        - progressbar2:      4.2.0
        - prometheus-client: 0.16.0
        - prometheus-flask-exporter: 0.22.4
        - protobuf:          3.19.6
        - psutil:            5.9.5
        - py:                1.11.0
        - pyarrow:           6.0.1
        - pyasn1:            0.5.0
        - pyasn1-modules:    0.3.0
        - pycodestyle:       2.10.0
        - pycparser:         2.21
        - pycryptodome:      3.17
        - pydantic:          1.10.7
        - pydocstyle:        6.3.0
        - pyflakes:          3.0.1
        - pygit2:            1.12.0
        - pygments:          2.15.1
        - pyjwt:             2.6.0
        - pyodbc:            4.0.39
        - pyparsing:         3.0.9
        - pyrsistent:        0.19.3
        - pysocks:           1.7.1
        - pytest:            7.3.1
        - pytest-cov:        3.0.0
        - pytest-forked:     1.6.0
        - pytest-mock:       3.10.0
        - pytest-xdist:      2.5.0
        - python-daemon:     3.0.1
        - python-dateutil:   2.8.2
        - python-dotenv:     0.20.0
        - python-nvd3:       0.15.0
        - python-slugify:    8.0.1
        - python-utils:      3.5.2
        - pytorch-lightning: 2.0.2
        - pytz:              2022.7.1
        - pytzdata:          2020.1
        - pyyaml:            6.0
        - querystring-parser: 1.2.4
        - regex:             2023.3.23
        - requests:          2.29.0
        - requests-oauthlib: 1.3.1
        - requests-toolbelt: 0.10.1
        - rfc3339-validator: 0.1.4
        - rich:              13.3.5
        - rsa:               4.9
        - s3fs:              2022.7.1
        - s3transfer:        0.5.2
        - scikit-learn:      1.2.1
        - scipy:             1.10.1
        - seaborn:           0.11.2
        - setproctitle:      1.3.2
        - setuptools:        67.7.2
        - setuptools-scm:    7.1.0
        - six:               1.16.0
        - sklearn:           0.0.post4
        - smart-open:        6.3.0
        - smmap:             5.0.0
        - sniffio:           1.3.0
        - snowballstemmer:   2.2.0
        - soupsieve:         2.4.1
        - spacy:             3.5.2
        - spacy-legacy:      3.0.12
        - spacy-loggers:     1.0.4
        - sqlalchemy:        1.4.47
        - sqlalchemy-jsonfield: 1.0.1.post0
        - sqlalchemy-stubs:  0.4
        - sqlalchemy-utils:  0.41.1
        - sqlparse:          0.4.4
        - srsly:             2.4.6
        - stevedore:         5.0.0
        - tabulate:          0.9.0
        - tenacity:          8.2.2
        - tensorboard:       2.11.2
        - tensorboard-data-server: 0.6.1
        - tensorboard-plugin-wit: 1.8.1
        - tensorflow:        2.11.1
        - tensorflow-estimator: 2.11.0
        - tensorflow-io-gcs-filesystem: 0.32.0
        - teradatasql:       17.20.0.19
        - termcolor:         1.1.0
        - text-unidecode:    1.3
        - thinc:             8.1.9
        - threadpoolctl:     3.1.0
        - thrift:            0.16.0
        - toml:              0.10.2
        - tomli:             2.0.1
        - tomlkit:           0.11.8
        - torch:             1.12.1
        - torchmetrics:      0.11.4
        - tornado:           6.3.1
        - tqdm:              4.65.0
        - typer:             0.7.0
        - types-awscrt:      0.16.16
        - types-cachetools:  5.3.0.5
        - types-pytz:        2021.3.8
        - types-pyyaml:      6.0.12.9
        - types-requests:    2.29.0.0
        - types-s3transfer:  0.6.0.post7
        - types-urllib3:     1.26.25.11
        - typing-extensions: 4.5.0
        - typing-inspect:    0.8.0
        - uc-micro-py:       1.0.1
        - unicodecsv:        0.14.1
        - urllib3:           1.26.15
        - virtualenv:        20.23.0
        - wasabi:            1.1.1
        - websocket-client:  1.5.1
        - werkzeug:          2.3.2
        - wheel:             0.40.0
        - wrapt:             1.15.0
        - wtforms:           3.0.1
        - yarl:              1.9.2
        - zipp:              3.15.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.13
        - version:           #1 SMP Thu Mar 23 09:54:12 UTC 2023

</details>
How you installed Lightning:  `pip`
Running in AWS EC2 environment.

cc @justusschock @awaelchli

**Comment by awaelchli** [comment]
Which file contains your code that you are running? Is it `hyperparameter_tune_lt_lstm.py`? Can you check the permissions on this file? Make sure you don't have any code in that file that changes the current working directory, otherwise the Trainer can't access and launch that script for the other processes. 

What happens if you run our debug script with `strategy="ddp"` and `devices=2`?
https://github.com/Lightning-AI/lightning/blob/master/examples/pytorch/bug_report/bug_report_model.py

**Comment by esaracin** [comment]
Yes, `hyperparameter_tune_lt_lstm.py` is the main script that invokes the training. For some more insight, it's a script that leverages Optuna to repeatedly train LSTMs with different hyper parameters, logging each to MLFLow, as it optimizes for maximum avg_prc. Ultimately, it instantiates successive `TrainLSTM` tasks (which are the classes with the code that creates the Trainer and runs `trainer.fit()`.). 

As an earlier test (suspecting the same thing), I chmod 777'd the script just to ensure it would be accessible to any other processes. I also attempted running it with `sudo` but was met with the same `No such file or directory` error.

I just ran the bug_report_script.py. For whatever reason, I had to change the `from lightning.pytorch import LightningModule, Trainer` statement to `from pytorch_lightning import LightningModule, Trainer` (since my system apparently doesn't know lightning by the former name). But with `strategy='ddp'` and `devices=2` (and the same Python environment), it runs without issue.

**Comment by esaracin** [comment]
Missed the forest for the trees: there **is** an error with the path it's trying to open the script from. There's one too many subdirectories of the same name. There should only be 2 subdirectories with the same name that leads to the script (there are 3 in the path the spawned processes are looking for the script in). 

So the question becomes: where do the subprocesses get their information on where the script is located? And why do they get it correct for `ddp_spawn` and not `ddp`.

Thanks for your continued help!

**Comment by awaelchli** [comment]
It is very simple. The regular ddp launcher creates a new process using subprocess.Popen:
https://github.com/Lightning-AI/lightning/blob/f6af74bf158a064673b5db284490b0da1f6c6852/src/lightning/pytorch/strategies/launchers/subprocess_script.py#L130-L134

And the command it uses gets constructed here in this function:
https://github.com/Lightning-AI/lightning/blob/984f49f7195ddc67e961c7c498ee6e19fc0cecb5/src/lightning/fabric/strategies/launchers/subprocess_script.py#L136C1-L142

As you can see, the path is basically just given by `os.path.abspath(sys.argv[0])`, which is just the name of the script and thus relying on the current working directory being the same from the start of the program. 

Since you are using a 3rd party framework that creates these tasks, I suspect that it is changing the working dir and so by the time the trainer launches, it can't find the file in the current dir. 

>  And why do they get it correct for ddp_spawn and not ddp?
Yes, very simple. ddp_spawn uses the `multiprocessing.spawn` from torch/Python to create new processes. So it makes sense that it doesn't fail this way. 

For single node training, ddp_spawn should serve you well. Did you see any limitations that pushed you to move away from it?

**Comment by esaracin** [comment]
Ahh-- I think this might be me more than the third party framework, then: I'm using a Makefile to invoke the script from several directories up. When I run from the correct directory, it appears to get past the initialization phase (thanks for your help figuring that out!).

Unfortunately, I think I've found a reason I should be using 'ddp_spawn' with my use-case: 'ddp' appears to start the entire script over again, while 'ddp_spawn' just kicks off the multiprocessing from the point of the beginning of training to the end of the script. There's a lot of data to load in and preparation at the beginning of the script, which I then have to wait for 4 processes to get through before training can begin with 'ddp'. With 'ddp_spawn', the threads exist only from the trainer.fit() call through the end of the script.

As for the issues I've had with 'ddp_spawn', see below...
_______________

I'm running a script that iteratively runs my training loop. With `ddp_spawn`, I can get through one training run just fine (and with all of the apparent benefits of multi-gpu training). However, when the next iteration comes along, I get met with this error in `spawn.py`:
<img width="1375" alt="Screenshot 2023-04-29 at 10 42 47 AM" src="https://user-images.githubusercontent.com/25374730/235330143-71f87a75-4339-4e46-8d68-e8aa6b005510.png">

So, running the same code, but on the next iteration, training fails to kick off correctly during the instantiation of those subprocesses, and I'm again met with a hanging process I have to kill. Effectively, it defeats the purpose of my loop, since only the first iteration works correctly, every time.

**Comment by awaelchli** [comment]
Hey @esaracin 
Is this still a problem or did you make further progress debugging the issues?
